openwisp_network_topology/consumers.py

```diff
@@ -2,31 +2,35 @@
 
 from asgiref.sync import async_to_sync
 from channels.generic.websocket import WebsocketConsumer
+from swapper import load_model
+
+Topology = load_model('topology', 'Topology')
 
 
 class TopologyConsumer(WebsocketConsumer):
     channel_layer_group = 'topology'
 
-    def _is_user_authenticated(self):
-        try:
-            assert self.scope['user'].is_authenticated is True
-        except (KeyError, AssertionError):
-            self.close()
-            return False
-        else:
+    def _is_user_authorized_to_view_topology(self, user, topology_pk):
+        if user.is_authenticated and user.is_superuser:
             return True
+        topology = Topology.objects.get(pk=topology_pk)
+        return (
+            user.is_authenticated
+            and user.has_perm(f'{Topology._meta.app_label}.view_topology')
+            and user.is_manager(topology.organization.pk)
+        )
```

Comment:
something I had forgot before is that there's a setting which can be changed to allow to view the topology data publicly, so we have to take that into account too.

I'd rewrite this as something like the following, please keep in mind I am writing this code from the top of my head so please double check this:

```python
if not user.is_authenticated and settings.API_AUTH_REQUIRED:
    return False
topology = Topology.objects.get(pk=topology_pk)
return (
    user.is_superuser or (
        user.is_manager(topology.organization_id) and
        user.has_perm(f'{Topology._meta.app_label}.view_topology')
    )
)
```